// arduino-ide-extension/lib/common/nls.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstallManually = exports.Retired = exports.Recommended = exports.Contributed = exports.Partner = exports.Type = exports.All = exports.Updatable = exports.Later = exports.Unknown = void 0;
const nls_1 = require("@theia/core/lib/common/nls");
exports.Unknown = nls_1.nls.localize('arduino/common/unknown', 'Unknown');
exports.Later = nls_1.nls.localize('arduino/common/later', 'Later');
exports.Updatable = nls_1.nls.localize('arduino/common/updateable', 'Updatable');
exports.All = nls_1.nls.localize('arduino/common/all', 'All');
exports.Type = nls_1.nls.localize('arduino/common/type', 'Type');
exports.Partner = nls_1.nls.localize('arduino/common/partner', 'Partner');
exports.Contributed = nls_1.nls.localize('arduino/common/contributed', 'Contributed');
exports.Recommended = nls_1.nls.localize('arduino/common/recommended', 'Recommended');
exports.Retired = nls_1.nls.localize('arduino/common/retired', 'Retired');
exports.InstallManually = nls_1.nls.localize('arduino/common/installManually', 'Install Manually');
