// arduino-ide-extension/lib/browser/contributions/sketchbook.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sketchbook = void 0;
const inversify_1 = require("@theia/core/shared/inversify");
const arduino_menus_1 = require("../menu/arduino-menus");
const examples_1 = require("./examples");
const protocol_1 = require("../../common/protocol");
const open_sketch_1 = require("./open-sketch");
const nls_1 = require("@theia/core/lib/common/nls");
let Sketchbook = class Sketchbook extends examples_1.Examples {
    onStart() {
        this.sketchServiceClient.onSketchbookDidChange(() => this.update());
    }
    async onReady() {
        this.update();
    }
    update() {
        this.sketchService.getSketches({}).then((container) => {
            this.register(container);
            this.menuManager.update();
        });
    }
    registerMenus(registry) {
        registry.registerSubmenu(arduino_menus_1.ArduinoMenus.FILE__SKETCHBOOK_SUBMENU, nls_1.nls.localize('arduino/sketch/sketchbook', 'Sketchbook'), { order: '3' });
    }
    register(container) {
        this.toDispose.dispose();
        this.registerRecursively([...container.children, ...container.sketches], arduino_menus_1.ArduinoMenus.FILE__SKETCHBOOK_SUBMENU, this.toDispose);
    }
    createHandler(uri) {
        return {
            execute: async () => {
                try {
                    await this.commandService.executeCommand(open_sketch_1.OpenSketch.Commands.OPEN_SKETCH.id, uri);
                }
                catch (err) {
                    if (protocol_1.SketchesError.NotFound.is(err)) {
                        // Force update the menu items to remove the absent sketch.
                        this.update();
                    }
                    else {
                        throw err;
                    }
                }
            },
        };
    }
};
Sketchbook = __decorate([
    (0, inversify_1.injectable)()
], Sketchbook);
exports.Sketchbook = Sketchbook;
