// arduino-ide-extension/lib/browser/theia/markers/problem-contribution.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProblemContribution = void 0;
const inversify_1 = require("@theia/core/shared/inversify");
const problem_contribution_1 = require("@theia/markers/lib/browser/problem/problem-contribution");
let ProblemContribution = class ProblemContribution extends problem_contribution_1.ProblemContribution {
    async initializeLayout(app) {
        // NOOP
    }
    setStatusBarElement(problemStat) {
        // NOOP
    }
    registerKeybindings(keybindings) {
        if (this.toggleCommand) {
            keybindings.registerKeybinding({
                command: this.toggleCommand.id,
                keybinding: 'ctrlcmd+alt+shift+m',
            });
        }
    }
};
ProblemContribution = __decorate([
    (0, inversify_1.injectable)()
], ProblemContribution);
exports.ProblemContribution = ProblemContribution;
